### ISSUE TITLE
Some questions

### ISSUE THREAD

**Issue by Khoa-NT** [mahmoodlab/HIPT#6]
Hi, thank you for a great paper.
I would like to ask some questions. 

1/ About the training code for [Slide-Level Classification](https://github.com/mahmoodlab/HIPT#weakly-supervised-training--evaluation). 
Are you guys still updating the code? Because I can't find the file [2-Weakly-Supervised-Train-Val/Model%20Walkthrough.ipynb](https://github.com/707884/707884/blob/master/2-Weakly-Supervised-Train-Val/Model%20Walkthrough.ipynb). Or `main.py` in the [training commands](https://github.com/mahmoodlab/HIPT#weakly-supervised-training--evaluation).
I only find the training in this [eval_linear.py](https://github.com/mahmoodlab/HIPT/blob/master/1-Hierarchical-Pretraining/eval_linear.py). Is it the training code for Slide-Level Classification? 

2/ You set `num_classes=0` in this [code](https://github.com/mahmoodlab/HIPT/blob/master/1-Hierarchical-Pretraining/eval_linear.py#L40), so I guess you train the $ViT_{WSI}-4096$ + an additional `classification layer` together at the Fine-tuning state. But I can't find the details of the `classification layer` for MIL in the paper or the supplementary. 
You only mentioned in the Fine-tuning part of the paper, that the  $ViT_{WSI}-4096$ is finetuned but no information about the classification layer. 
In addition, in this [code](https://github.com/mahmoodlab/HIPT/blob/master/1-Hierarchical-Pretraining/eval_linear.py#L173), you only train one Linear layer for the `classification layer` from the feature. 
I'm confused. I thought you train the $ViT_{WSI}-4096$ + `classification layer`.
Can you give me more details about the Fine-tuning state?

3/ And also, does `main_dino256.py` in the [command](https://github.com/mahmoodlab/HIPT#hierarchical-pretraining-for-vit-16256-models--pretrained-models) is the [1-Hierarchical-Pretraining/main_dino.py](https://github.com/mahmoodlab/HIPT/blob/master/1-Hierarchical-Pretraining/main_dino.py)?

4/ In the Supplementary, Table 4: 
![image](https://user-images.githubusercontent.com/12138158/177123203-cb3d2925-97f8-4ca0-b57b-8dacc0724871.png)
In the 4th case, Is it a typo? 
* If you used only the $ViT-16_{PF}$, $ViT-256_{PF}$ then an `AP-4096` should be at the red line.  

* If it is not the typo, then how you can get the result with only 2 $ViT-16_{PF}$, $ViT-256_{PF}$?
 Furthermore, It means using only 2 pretrained $ViT-16_{PF}$, $ViT-256_{PF}$ from DINO is the best result ( the same as the best result in Table 1)? Then it means $ViT_{WSI}-4096$ is not important, right?
 
 Thank you for supporting me.

**Comment by Richarizardd** [comment]
Hi @Khoa-NT - Thank you for your interest in this repository

1. We uploaded the training scaffold code to [./2-Weakly-Supervised-Subtyping/](https://github.com/mahmoodlab/HIPT/tree/master/2-Weakly-Supervised-Subtyping). Most of the training scaffold code is heavily borrowed from [CLAM](https://github.com/mahmoodlab/CLAM). For the survival code ([MCAT](https://github.com/mahmoodlab/MCAT)) and future works that we release that continuing to build off of CLAM, we will just point to the original scaffold code w/ updated models that we have added.

Model walkthrough code is found here for [3-stage HIPT.](https://github.com/mahmoodlab/HIPT/blob/master/2-Weakly-Supervised-Subtyping/Model%20Walkthrough.ipynb)

The [eval_linear](https://github.com/mahmoodlab/HIPT/blob/master/1-Hierarchical-Pretraining/eval_linear.py) was from the original [DINO](https://github.com/facebookresearch/dino) repository, which we did not touch.

2. As hopefully made transparent in the [model walkthrough code](https://github.com/mahmoodlab/HIPT/blob/master/2-Weakly-Supervised-Subtyping/Model%20Walkthrough.ipynb), the 3-Stage HIPT only trains the last vision transformer stage (for aggregating $x_{4096}$  features). The input into this HIPT model uses pre-extracted $x_{256}$ features, but unlike ```[M_256 x D]```  tensor bag-like inputs used in CLAM and most other models (where $M_{256}$ is the number of 256-sized patches in the WSI) , the input shape in this setting is ```[M_4K x 256 x D]```, where $M_{4K}$ is the number of 4K-sized patches and 256 is the number of 256-sized patches in a single 4K-sized patch. To create ```[M_4K x 256 x D]```, for the  $M_{4K} x_{4096}$ images in your WSI, you can extract ```[1 x 256 x D]``` features from each ```x_{4096}``` image using Two-Stage HIPT with loaded weights, then stack these tensors along the zeroth dimension.

To point something else out, since the main HIPT method we evaluate uses pretrained + pre-extracted features at the lower level, we can think of 3-Stage-HIPT as just doing MIL with pretrained 2-Stage-HIPT features (of 4K resolution). However, still using pre-extracted $x_{256}$ features was important for doing ablation experiments in 3-Stage-HIPT without pretraining. Lots of fun variations to try!

4. Will standardize the naming convention of the scripts.

5. This may be a typo - it should say ViT-4096. Will update the arXiv soon

**Comment by invoker-LL** [comment]
> Hi @Khoa-NT - Thank you for your interest in this repository
> 
> 1. We uploaded the training scaffold code to [./2-Weakly-Supervised-Subtyping/](https://github.com/mahmoodlab/HIPT/tree/master/2-Weakly-Supervised-Subtyping). Most of the training scaffold code is heavily borrowed from [CLAM](https://github.com/mahmoodlab/CLAM). For the survival code ([MCAT](https://github.com/mahmoodlab/MCAT)) and future works that we release that continuing to build off of CLAM, we will just point to the original scaffold code w/ updated models that we have added.
> 
> Model walkthrough code is found here for [3-stage HIPT.](https://github.com/mahmoodlab/HIPT/blob/master/2-Weakly-Supervised-Subtyping/Model%20Walkthrough.ipynb)
> 
> The [eval_linear](https://github.com/mahmoodlab/HIPT/blob/master/1-Hierarchical-Pretraining/eval_linear.py) was from the original [DINO](https://github.com/facebookresearch/dino) repository, which we did not touch.
> 
> 2. As hopefully made transparent in the [model walkthrough code](https://github.com/mahmoodlab/HIPT/blob/master/2-Weakly-Supervised-Subtyping/Model%20Walkthrough.ipynb), the 3-Stage HIPT only trains the last vision transformer stage (for aggregating x4096  features). The input into this HIPT model uses pre-extracted x256 features, but unlike `[M_256 x D]`  tensor bag-like inputs used in CLAM and most other models (where M256 is the number of 256-sized patches in the WSI) , the input shape in this setting is `[M_4K x 256 x D]`, where M4K is the number of 4K-sized patches and 256 is the number of 256-sized patches in a single 4K-sized patch. To create `[M_4K x 256 x D]`, for the  M4Kx4096 images in your WSI, you can extract `[1 x 256 x D]` features from each `x_{4096}` image using Two-Stage HIPT with loaded weights, then stack these tensors along the zeroth dimension.
> 
> To point something else out, since the main HIPT method we evaluate uses pretrained + pre-extracted features at the lower level, we can think of 3-Stage-HIPT as just doing MIL with pretrained 2-Stage-HIPT features (of 4K resolution). However, still using pre-extracted x256 features was important for doing ablation experiments in 3-Stage-HIPT without pretraining. Lots of fun variations to try!
> 
> 4. Will standardize the naming convention of the scripts.
> 5. This may be a typo - it should say ViT-4096. Will update the arXiv soon

Can you share more details about the subtyping training or how to  modify CLAM to adapt the HIPT features instead of RESNET-50(layer3's pooling, N*1024 dim)? Especially the beginning linear layers with 1024->512->256 dim in CLAM for HIPT features N*384? Also I find that the there is also no train/evaluation result with traditional resnet-50 + CLAM baseline.

**Comment by Richarizardd** [comment]
Hi @invoker-LL - given the breadth of the experiments already in the paper, we did not have time to run ablation experiments with other feature embedding types. There are many permutations of experiments to do going down that path (not only ResNet-50 evaluation for every model and for cancer type, also doing Hierarchical Pretraining of ResNet-50 features). The focus of this paper is primarily looking at architecture + hierarchical pretraining. We are saving those other experiments for follow-up work.

You can see the [modification here] for CLAM using VIT-16 features (of 384-dim).(https://github.com/mahmoodlab/HIPT/blob/57a5c4b41cad47c6094908dc745752fae8fb17c4/2-Weakly-Supervised-Subtyping/models/model_clam.py#L23)

**Comment by Richarizardd** [comment]
[See these files](https://github.com/mahmoodlab/HIPT/tree/master/2-Weakly-Supervised-Subtyping/dataset_csv), which contains the total list of WSIs evaluated. Though there are more patients included in the ```10foldcv_subtype/tcga_brca/split_{i}.csv``` split, patients with missing WSIs get excluded [when slicing the dataframe](https://github.com/mahmoodlab/HIPT/blob/57a5c4b41cad47c6094908dc745752fae8fb17c4/2-Weakly-Supervised-Subtyping/datasets/dataset_generic.py#L198).

I should update the split csv files to clarify this confusion.

**Comment by Khoa-NT** [comment]
Hi, I would like to ask some more questions about training classification.

The code you provided for training the classification task. E.g., the `tcga_brca_subtype`
<details>
<summary>
Full List of Training Classification Commands
</summary>

```python
GPU=0
DATAROOT=/path/to/TCGA_ROOT_DIR/
TASK=tcga_brca_subtype
CUDA_VISIBLE_DEVICES=$GPU python main.py --data_root_dir $DATAROOT --model_type hipt_lgp --task $TASK --prop 0.25
CUDA_VISIBLE_DEVICES=$GPU python main.py --data_root_dir $DATAROOT --model_type hipt_lgp --task $TASK --prop 1.0
CUDA_VISIBLE_DEVICES=$GPU python main.py --data_root_dir $DATAROOT --model_type hipt_lgp --task $TASK --prop 0.25 --pretrain_4k vit4k_xs_dino
CUDA_VISIBLE_DEVICES=$GPU python main.py --data_root_dir $DATAROOT --model_type hipt_lgp --task $TASK --prop 1.0 --pretrain_4k vit4k_xs_dino
CUDA_VISIBLE_DEVICES=$GPU python main.py --data_root_dir $DATAROOT --model_type hipt_lgp --task $TASK --prop 0.25 --pretrain_4k vit4k_xs_dino --freeze_4k
CUDA_VISIBLE_DEVICES=$GPU python main.py --data_root_dir $DATAROOT --model_type hipt_lgp --task $TASK --prop 1.0 --pretrain_4k vit4k_xs_dino --freeze_4k
```
</details>

1/ Following this [code](https://github.com/mahmoodlab/HIPT/blob/master/2-Weakly-Supervised-Subtyping/models/model_hierarchical_mil.py#L72), the upper commands represent the below experiments: 
* $ViT-16_{PF}, ViT-256, ViT-4096$ with 25% and 100% training
* $ViT-16_{PF}, ViT-256_{P}, ViT-4096$ with 25% and 100% training
* $ViT-16_{PF}, ViT-256_{PF}, ViT-4096$ with 25% and 100% training

Is it right?

2/ Then the code [HIPT_None_FC](https://github.com/mahmoodlab/HIPT/blob/master/2-Weakly-Supervised-Subtyping/models/model_hierarchical_mil.py#L23) is used for the case $ViT-16_{PF}, AP-256, AP-4096$, right?
And the commands to run are:
```python
CUDA_VISIBLE_DEVICES=$GPU python main.py --data_root_dir $DATAROOT --model_type hipt_n --task $TASK --prop 0.25
CUDA_VISIBLE_DEVICES=$GPU python main.py --data_root_dir $DATAROOT --model_type hipt_n--task $TASK --prop 1.0
```
Am I correct? 


3/ I can't find the definition of `pretrain_WSI` in this [code](https://github.com/mahmoodlab/HIPT/blob/master/2-Weakly-Supervised-Subtyping/models/model_hierarchical_mil.py#L979). 
And also there is no definition of `self.global_vit` in this [code](https://github.com/mahmoodlab/HIPT/blob/master/2-Weakly-Supervised-Subtyping/models/model_hierarchical_mil.py#L140).
If I understand correctly, `self.global_vit` is the $ViT_{WSI}-4096$, right?

4/ Following the [Model Walkthrough.ipynb](https://github.com/mahmoodlab/HIPT/blob/master/2-Weakly-Supervised-Subtyping/Model%20Walkthrough.ipynb). I can assume that `pretrain_WSI='None'`.
Then, it means the stage-3 $ViT_{WSI}-4096$ is not a Vision transformer. It's simply just a stack of `TransformerEncoderLayer`.
Is it correct? or You are still updating the code for finetuning the $ViT_{WSI}-4096$?

Thank you for supporting me.

**Comment by juanigp** [comment]
Hi @Khoa-NT @Richarizardd , I would also like to know about the normalization transform.

**Comment by Richarizardd** [comment]
Hi @juanigp @Khoa-NT - The normalization transform is ```mean = (0.5,0.5,0.5)``` &```std = (0.5,0.5,0.5)```. I may have accidentally pushed some code that I was playing around with a few months ago (since reverted back). My apologies.
